### ISSUE TITLE
[BUG] Inaccurate code coverage

### ISSUE THREAD

**Issue by edumserrano** [cenfun/monocart-reporter#29]
Hi @cenfun,

When instrumenting playwright tests for an angular app using `istanbul` via the `babel-plugin-istanbul plugin` the code coverage shows accurately whilst when using `V8 + monocart-reporter` some lines that should be covered show as uncovered even though there's an indication those lines have been executed X number of times.

The images below show a part of the code coverage for a file which show the problem.

As per our conversation on `Playwright` discord channel I've created a repo which allows you to easily reproduce the problem. Please see [edumserrano/monocart-code-coverage-demo](https://github.com/edumserrano/monocart-code-coverage-demo).

- Istanbul code coverage report:

![istanbul](https://github.com/cenfun/monocart-reporter/assets/15857357/686e14d1-7e69-4474-8610-319b1ed3ecf9)

- Monocart code coverage report:

![monocart](https://github.com/cenfun/monocart-reporter/assets/15857357/4a094674-661d-4f91-9e0e-54cb86189b82)

**Comment by cenfun** [comment]
It should be fixed, please try new version monocart-reporter@1.6.29

The root cause is that the original location cannot be found from the sourcemap sometimes.

Sourcemap matching is always a problem, especially when some non-native js syntax is used.
For example, the following problem still cannot be solved because there is a wrong match in the sourcemap:

![image](https://github.com/cenfun/monocart-reporter/assets/334577/c1729f87-4275-48fb-93d1-df604b692dcc)

we can find out the mismatch with tool: https://evanw.github.io/source-map-visualization/
![image](https://github.com/cenfun/monocart-reporter/assets/334577/41c7d932-86cd-44aa-a129-dbce2b43c4cf)

It seems that sourcemap related issues should be traced back to compilation tools like babel.

Thanks for such a detailed description of the problem, more suggestions are welcome.

**Comment by cenfun** [comment]
please update the config, the previous `sourceFilter` is not right
![image](https://github.com/cenfun/monocart-reporter/assets/334577/ef2b99af-8268-40c4-9be1-ad2cec00afd0)

**Comment by edumserrano** [comment]
Awesome! All is working 🎉  and you made it even better by not having the `webpack:/` at the start so now it starts with just `src` 👏 

I do have a followed question. In the coverage details fly out could we also remove the `webpack` mention from there? I'd assume it would be related with the fix you did to remove `webpack` mention from the summary tree ?

![image](https://github.com/cenfun/monocart-reporter/assets/15857357/4522724e-5d6d-4a17-b7e9-650c4b4e611e)

**Comment by cenfun** [comment]
no problem. 
actually, the `webpack:/xxx` is the url of the source file, and it will be converted to the path of the source file, 
If we don't like url but prefer path, we can use path instead of url next time
![image](https://github.com/cenfun/monocart-reporter/assets/334577/10b87718-3926-4986-96d9-19bbcc543c17)

**Comment by edumserrano** [comment]
I can't be sure if that change makes sense for all cases but I think it does at least for this scenario. I'd like to have it without `webpack`, I do think using the sourcePath` makes more sense but I'll leave that up for you to decide.

To help others, I'll create a branch on the [demo repo](https://github.com/edumserrano/monocart-code-coverage-demo) that contains the fix and update the README to mention the fix is on the branch:
- update to `1.6.29`
- update the `playwright.config.ts`

I think you can close this issue whenever you want @cenfun . Thank you for sorting this out and so quickly 🥇

**Comment by cenfun** [comment]
just removed the `webpack:///` see commit: https://github.com/cenfun/monocart-reporter/commit/1e9542cb945a72e95af4b3752148944fb38661ba
I prefer to have it without `webpack:///` too. 

I understand that you spend a lot of time on the code coverage, so do I. 
But currently few users use this code coverage feature, suggestions are welcome.
Thanks

**Comment by edumserrano** [comment]
This might be a stupid question but the change you mentioned from commit 1e9542c  will be in a new version right? `1.6.30` ?
Just checking cause I've been refreshing https://www.npmjs.com/package/monocart-reporter waiting for the new version 😆

**Comment by cenfun** [comment]
yes, `1.6.30` will be released when few minor issues are fixed. sorry for the later.

**Comment by cenfun** [comment]
`1.6.30` have released. please have a try.

**Comment by edumserrano** [comment]
I can confirm version `1.6.30` is working as expected. All looking good 👏

**Comment by edumserrano** [comment]
@cenfun found some more examples where code coverage from `V8+monocart-reporter` is incorrect when comparing with coverage using `babel+istanbul`. Perhaps we should reopen this issue.

I've had to expand the code of the demo repo and created a branch named `temp` [edumserrano/monocart-code-coverage-demo/tree/temp](https://github.com/edumserrano/monocart-code-coverage-demo/tree/temp).

Following again the instructions on the README:
- if you run `npm run test-monocart` and then `npm run open-monocart-coverage` you will see the code coverage with `V8+monocart-reporter`. 
- if you run `npm run test-istanbul` and then `npm run open-istanbul-coverage` you will see the code coverage with `babel+istanbul`.

You can see a difference in for instance:
- `src/app/features/find-institution/find-institution.view-model.ts`
- `src/app/features/find-institution/recently-used/recently-used-institutions.service.ts`

In both of the above files, the `babel+istanbul` shows correct code coverage as opposed to the `V8+monocart-reporter` version.  See images below:

<details>
  <summary>babel+istanbul find-institution.view-model.ts</summary>
  
![istanbul-find-insitution-viewmodel](https://github.com/cenfun/monocart-reporter/assets/15857357/b8d78878-847b-4447-aeda-e9a207960294)
</details>

<details>
  <summary>V8+monocart-reporter find-institution.view-model.ts</summary>

**The return statement that shows as uncovered should show as covered.**

![monocart-reporter-find-institution-viewmodel](https://github.com/cenfun/monocart-reporter/assets/15857357/fd5193d5-c204-4272-9d70-b1208694a230)
</details>

<details>
  <summary>babel+istanbul recently-used-institutions.service.ts</summary>
  
![istanbul-recently-used-service](https://github.com/cenfun/monocart-reporter/assets/15857357/95664974-d70d-4aa5-8b19-45809c73ef4b)
</details>

<details>
  <summary>V8+monocart-reporter recently-used-institutions.service.ts</summary>

**The `moveToFront()` function and the `if/else` branch above show as uncovered in this image should show as covered.**

![image](https://github.com/cenfun/monocart-reporter/assets/15857357/420b0510-34ef-4ad9-9d78-d12743077978)
</details>

Another example that shows incorrect coverage is on an `html` file. For this I can't compare with `babel+istanbul` because that setup is only capturing code coverage for TypeScript files. However, the image below from the `V8+monocart-reporter` setup is for the `src/app/features/find-institution/institutions-list/institutions-list.component.html` file and it shows as such:

![monocart-reporter-institutions-list-component](https://github.com/cenfun/monocart-reporter/assets/15857357/c2e44634-a210-4e75-b5ac-e8623a02f16d)

Both of the above red lines should show as covered. And, similarly to the original bug reported on this issue, notice that one of the lines that shows as uncovered as the indication that it has been executed `5x`.

> **Note**
> 1) there are many tests now in this branch, it make take a few seconds or a minute to complete.
> 2) many are screenshot tests which will only work the second time you run then because I haven't pushed the screenshots to the repo. And even if you rerun there are 2 tests that will always fail (that is expected).
> 3) this is a temp branch to show the problem and it contains an expanded version of the code in `main` branch. However I'm not comfortable with leaving this branch available forever so I will be deleting it once you're able to replicate and diagnose the issue.

**Comment by cenfun** [comment]
##  Let's check if v8 coverage is correct in dist file first
-  set `excludeDistFile: false` and run test, the dist file `main.js` will be in report
![image](https://github.com/cenfun/monocart-reporter/assets/334577/d10570f6-0b1a-446b-ba43-49c63acbd569)
![image](https://github.com/cenfun/monocart-reporter/assets/334577/77a4e37f-2a08-4b4f-a16f-25accfbf30d9)

- find out what the line is in dist file for line `102` in `find-institution.view-model.ts` (uncovered)
![image](https://github.com/cenfun/monocart-reporter/assets/334577/205a4807-6013-4e7a-ba56-b9c2e7671fdb)
download `http://localhost:4999/main.js` and `http://localhost:4999/main.js.map` when server is startup.
open them in the [source-map-visualization](https://evanw.github.io/source-map-visualization/) and find out the line is `11135` in dist file:
![image](https://github.com/cenfun/monocart-reporter/assets/334577/96d26a84-59a8-4f1e-84a6-e25451ac9b34)

-  check the line `11135` and it is uncovered, so the v8 coverage should be correct.
![image](https://github.com/cenfun/monocart-reporter/assets/334577/301d8640-4490-4132-b6fd-e5ca51e2066f)

## I don't think we can directly compare Istanbul and V8 coverage like you do.
- the runtime code is different even the source code is the same, but the dist file is different 
- the tests are not executed exactly the same, for example, playwright execute the tests in parallel, test1 may be executed before or after test2
- whether the page is reloaded during testing? coverage data may need to be merged. I don't know the details but it looks complicated.

Is it possible that Istanbul is wrong and V8 is correct? maybe need more verification.

**Comment by cenfun** [comment]
confirmed it is a bug. the uncovered lines have not mered with covered lines some times. thanks for the example provided.

**Comment by cenfun** [comment]
It's so hard to find the root cause. 
First, upgrage to monocart-reporter@1.6.34, I refactored the coverage related codes. 
But still needs some changes to auto fixtures, adding one more option `resetOnNavigation: false`
```js
await Promise.all([
    page.coverage.startJSCoverage({
        resetOnNavigation: false
    }),
    page.coverage.startCSSCoverage({
        resetOnNavigation: false
    }),
]);
```
This is because there are multiple pages in one test, the coverage data will be reset if navigation by default
```js
test(`recently used @find-institution @recently-used`, async ({ page }) => {
    const findInstitutionPage = new FindInstitutionPage(page);
    const chooseSignInPage = new ChooseSignInPage(page);
```
see [resetOnNavigation](https://playwright.dev/docs/api/class-coverage#coverage-start-js-coverage-option-reset-on-navigation)

![image](https://github.com/cenfun/monocart-reporter/assets/334577/8cacd1c8-540d-4a20-bf1d-461962d9fcf2)

**Comment by edumserrano** [comment]
The issue on the HTML has been fixed with `1.6.34`

Before
![image](https://github.com/cenfun/monocart-reporter/assets/15857357/61cb9c97-666c-4f06-819d-872157e8f1f8)

After
![image](https://github.com/cenfun/monocart-reporter/assets/15857357/187ca827-7c15-4c2b-9f32-1268d316aa69)
